source/common/upstream/cluster_manager_impl.cc

```diff
@@ -835,38 +836,45 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainConnPools(const Hos
 
 void ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainConnPools(
     HostSharedPtr old_host, ConnPoolsContainer& container) {
-  container.drains_remaining_ += container.pools_.size();
-
-  for (const auto& pair : container.pools_) {
-    pair.second->addDrainedCallback([this, old_host]() -> void {
-      if (destroying_) {
-        // It is possible for a connection pool to fire drain callbacks during destruction. Instead
-        // of checking if old_host actually exists in the map, it's clearer and cleaner to keep
-        // track of destruction as a separate state and check for it here. This also allows us to
-        // do this check here versus inside every different connection pool implementation.
-        return;
-      }
+  container.drains_remaining_ += container.pools_->size();
+
+  // make a copy to protect against the erase in the callback
+  auto pools = container.pools_;
+  pools->addDrainedCallback([this, old_host]() -> void {
+    if (destroying_) {
+      // It is possible for a connection pool to fire drain callbacks during destruction. Instead
+      // of checking if old_host actually exists in the map, it's clearer and cleaner to keep
+      // track of destruction as a separate state and check for it here. This also allows us to
+      // do this check here versus inside every different connection pool implementation.
+      return;
+    }
 
-      ConnPoolsContainer& container = host_http_conn_pool_map_[old_host];
-      ASSERT(container.drains_remaining_ > 0);
-      container.drains_remaining_--;
-      if (container.drains_remaining_ == 0) {
-        for (auto& pair : container.pools_) {
-          thread_local_dispatcher_.deferredDelete(std::move(pair.second));
-        }
-        host_http_conn_pool_map_.erase(old_host);
-      }
-    });
+    ConnPoolsContainer* to_clear = getHttpConnPoolsContainer(old_host);
+    if (to_clear == nullptr) {
+      // This could happen if we have cleaned out the host before iterating through every connection
+      // pool. Handle it by just continuing
+      return;
+    }
 
-    // The above addDrainedCallback() drain completion callback might execute immediately. This can
-    // then effectively nuke 'container', which means we can't continue to loop on its contents
-    // (we're done here).
-    if (host_http_conn_pool_map_.count(old_host) == 0) {
-      break;
+    ASSERT(to_clear->drains_remaining_ > 0);
+    to_clear->drains_remaining_--;
+    if (to_clear->drains_remaining_ == 0 && to_clear->ready_to_drain_) {
+      clearContainer(old_host, *to_clear);
     }
+  });
+
+  container.ready_to_drain_ = true;
```

Comment:
What do others think of this? On the one hand, it's kind of ugly. On the other hand, the only way to truly solve it that I can see is to either "check" for this case in the map itself, or expose an iterator from the map. In the first case, we're guessing at the intentions of the callback -- i.e. what if the map is modified in a way other than "clearing" it? It may be difficult to detect this case. In the second, we're removing one of the benefits of having the map abstract the underlying storage. 

I think this is just the unfortunate situation caused by the callback referring to its owner, and it being invoked immediately. So, it's a bunch of tradeoffs. I'd personally rather keep the map pure, and deal with this hopefully rare case in the callback itself.

Reply:
Yeah I think this is OK, maybe a comment somewhere about why we have this?